#'Differential abundance analysis
#'
#'This function finds the features that are significantly differentially abundant in the provided data, 
#'using DESeq implementation which models taxa abundance as a negative binomial distribution. See \link[DESeq2]{DESeq}  for more details. The significance 
#'of differentially abundant taxa is defined by log2 fold change and pvalue thresholds. These features are 
#'then assigned importance using random forest classifer. The measure of importance used in this implementation
#'is mean decrease in accuracy. 
#'
#' @param physeq (Required). A link[phyloseq]{phyloseq} object containing merged information of abundance,
#'        taxonomic assignment, sample data including the measured variables and categorical information
#'        of the samples, and / or phylogenetic tree if available.
#' @param grouping_column (Required). Character string specifying name of a categorical variable that is preffered for grouping the.
#'        information, this should be one of the components of grouping vector.
#' @param output_norm (optional). A character string specifying method to be used for transforming abundance data to be used for plotting.
#'        note that, this normalisation occurs after DESeq analysis. Therefore, it is strictly for purposes of the output data
#'        and not for differential expression analysis.
#' @return Returns a ggplot object. This can further be manipulated as preferred by user.
#' @param  pvalue.threshold. Cut off p-value for significance of differentially abundance taxa, default is 0.05.
#' @param  lfc.threshold. Threshold for log2 fold change over which significance of differentially expressed taxa is considered.
#' @return Returns a list of three items: \itemize{
#'         \item  SignfeaturesTable: A \code{data.frame} of taxa with raw and adjusted pvalues, basemean, log2 fold change and significance labels. 
#'         computed by using raw p-values.
#'         \item  importance: A \code{data.frame} of taxa mean decrease in accuracy as obtained by random forest classifier.
#'         \item plotdata: A \code{data.frame} of taxa and corresponding corrected p-values, importance rank organised 
#'         in form accepted by ggplot. 
#'        }
#' @examples 
#' data(pitlatrine)
#' physeq<-data(physeq)
#' deseq_sig  <- differential_abundance(physeq, grouping_column = "Country")
#' 
#' @references \url{http://userweb.eng.gla.ac.uk/umer.ijaz/}, Umer Ijaz, 2015
#' 
#' @author Alfred Ssekagiri \email{assekagiri@gmail.com},  Umer Zeeshan Ijaz \email{Umer.Ijaz@glasgow.ac.uk}
#' 
#' @import phyloseq
#' @import DESeq2
#'
#' @export differential_abundance
#' 
differential_abundance <- function(physeq, grouping_column,pvalue.threshold=0.05,lfc.threshold=0, 
                          filename = NULL,output_norm="log-relative"){
  
  abund_table <- otu_table(physeq)
  meta_table <-data.frame(sample_data(physeq))
  
  #==get count data to be used in deseq ==========#
  countData = round(as(abund_table, "matrix"), digits = 0)+1
  if(!taxa_are_rows(physeq)){
    countData = t(countData)
  }
  #==add a dummy column corresponding to grouping variable ==#
  meta_table$Groups <- meta_table[,grouping_column]
  #== create deseq compatible matrix and then test differential abundance in the groups=#
  dds <- DESeqDataSetFromMatrix(countData, meta_table, as.formula(~Groups))
  data_deseq_test = DESeq(dds)
  #==Extract results of the test =======#
  res = results(data_deseq_test, cooksCutoff = FALSE)
  res_tax = cbind(as.data.frame(res), as.matrix(countData[rownames(res), ]), OTU = rownames(res))
  res_tax_sig = subset(res_tax, padj < pvalue.threshold & lfc.threshold < abs(log2FoldChange))
  res_tax$Significant <- ifelse(rownames(res_tax) %in% rownames(res_tax_sig) , "Yes", "No")
  res_tax$Significant[is.na(res_tax$Significant)] <- "No"
  
  sig_res <- res_tax[rownames(res_tax_sig),"OTU"] #names of significant features
  
  #==normalisation abundance data for random forest clasification and output ==# 
  if(is.null(output_norm)){
    physeq <- normalise_data(physeq,output_norm)
    data <- as.data.frame(otu_table(physeq))
  }
  else{
    data <- abund_table
  }
  
  #==get importance measure of significant features using random forest
  subset.data<-data.frame(data[,as.character(res_tax[rownames(res_tax_sig),"OTU"])])
  rf_res <- randomforest_res(subset.data, meta_table$Groups)
  df_accuracy <- rf_res$importance
  sig_res <- sig_res[sig_res %in% df_accuracy$Sample]
  
  #organise data for output
  df <- NULL
  for(i in sig_res){
    rank <- (subset(df_accuracy, df_accuracy$Sample==i))$rank
    rank_label <- paste(paste(rep("â¬›", rank),collapse=""),max(df_accuracy$rank)+1-rank)
    tmp<-data.frame(data[,i],meta_table$Groups, rep(paste(rank_label)), rep(paste(i," padj = ",sprintf("%.5g",res_tax[i,"padj"]),sep=""),dim(data)[1]))
    colnames(tmp)<-c("Value","Groups","Rank","Taxa") 
    if(is.null(df)){df<-tmp} else { df<-rbind(df,tmp)}
  }
  
  data_to_write=NULL
  if(!is.null(filename)){
    res_table <- data.frame(res_tax_sig$baseMean , res_tax_sig$log2FoldChange,res_tax_sig$padj)
    row.names(res_table)<- rownames(res_tax_sig)
    data_to_write<-res_tax_sig[,c("baseMean","log2FoldChange","pvalue","padj")]
    data_to_write$Upregulated<-levels(meta_table[,grouping_column])[as.numeric(data_to_write$log2FoldChange>0)+1]
    write.csv(data_to_write,paste("NB_significant_",paste(levels(meta_table$Groups),collapse="_vs_"),".csv",sep=""))
  }
  
  out_put <- list("SignFeaturesTable"=res_tax,"plotdata"=df,"importance"=df_accuracy)
  return(out_put)
}


